lib/rust/ensogl/core/src/application/frp.rs

```diff
@@ -506,3 +506,536 @@ macro_rules! define_endpoints_emit_alias {
+/// `<Foo:(Clone+Debug), Bar:(Debug+'static)>`.
+///
+/// The macro will create two sets of API: a public API and an private API. The public API is
+/// intended to be used to send events into the API and receive events from the API, while the
+/// private API is intended to process the events sent in through the public API and generate
+/// the events going out though the public API.
+///
+/// Please note that additional fields are added automatically to the API. In particular, an
+/// output `focused(bool)`, and inputs `focus()`, `defocus()`, and `set_focus(bool)` are always
+/// defined and connected. They are mainly used for the shortcut manager to send commands only
+/// to focused GUI elements.
+#[macro_export]
+macro_rules! define_endpoints_2 {
+    (
+        $([$($global_opts:tt)*])?
+        $(<$($param:ident $(:($($constraints:tt)*))?),*>)?
+
+        $(Input { $([$($input_opts:tt)*])?
+            $($(#[doc=$($in_doc:tt)*])*
+            $in_field : ident ($($in_field_type : tt)*)),* $(,)?
+        })?
+
+        $(Output { $([$($output_opts:tt)*])?
+            $($(#[doc=$($out_doc:tt)*])*
+            $out_field : ident ($($out_field_type : tt)*)),* $(,)?
+        })?
+    ) => {
+        $crate::define_endpoints_2! {
+            NORMALIZED
+            $([$($global_opts)*])?
+            $(<$($param $(:($($constraints)*))?),*>)?
+
+            Input { $($([$($input_opts)*])?)?
+                /// Focus the element. Focused elements are meant to receive shortcut events.
+                focus(),
+                /// Defocus the element. Non-focused elements are meant to be inactive and don't
+                /// receive shortcut events.
+                defocus(),
+                /// Wrapper for `focus` and `defocus`.
+                set_focus(bool),
+                $($($(#[doc=$($in_doc )*])*
+                $in_field ($($in_field_type )*)),*)?
+            }
+
+            Output { $($([$($output_opts)*])?)?
+                /// Focus state checker.
+                focused(bool),
+                $($($(#[doc=$($out_doc)*])*
+                $out_field ($($out_field_type)*)),*)?
+            }
+        }
+    };
+
+    (
+        NORMALIZED
+        $([$($global_opts:tt)*])?
+        $(<$($param:ident $(:($($constraints:tt)*))?),*>)?
+
+        Input { $([$($input_opts:tt)*])?
+            $($(#[doc=$($in_doc :tt)*])*
+            $in_field : ident ($($in_field_type : tt)*)),* $(,)?
+        }
+
+        Output { $([$($output_opts:tt)*])?
+            $($(#[doc=$($out_doc:tt)*])*
+            $out_field : ident ($($out_field_type : tt)*)),* $(,)?
+        }
+    ) => {
+        use $crate::frp::IntoParam;
+
+        /// Frp network and endpoints.
+        #[derive(Debug,Derivative)]
+        #[derive(CloneRef, Clone)]
+        #[allow(missing_docs)]
+        pub struct Frp $(<$($param $(:$($constraints)*)?),*>)? {
+            public: api::Public $(<$($param),*>)?,// deref
```

Comment:
what does `// deter` mean?

Reply:
This was a comment left from your example to indicate the struct should have a `Deref` implementation for `public`. Removed the comment.

Reply:
This comment was removed, but still left in other places.